compiler/rustc_metadata/src/locator.rs

```diff
@@ -798,8 +798,15 @@ fn get_metadata_section<'p>(
                 )));
             }
 
+            // Length of the compressed stream - this allows linkers to pad the section if they want
+            let usize_len = core::mem::size_of::<u32>();
+            let Ok(len_bytes) = <[u8; 4]>::try_from(&buf[header_len..cmp::min(header_len + usize_len, buf.len())]) else {
+                return Err(MetadataError::LoadFailure("invalid metadata length found".to_string()));
+            };
+            let compressed_len = u32::from_be_bytes(len_bytes) as usize;
+
             // Header is okay -> inflate the actual metadata
-            let compressed_bytes = &buf[header_len..];
+            let compressed_bytes = &buf[header_len..compressed_len + header_len];
```

Comment:
```suggestion
            let compressed_bytes = &buf[header_len..header_len + compressed_len];
```
I think this is a bit clearer.

Reply:
By the way this needs to add another 4 bytes on both sides of the range to cover the length field.